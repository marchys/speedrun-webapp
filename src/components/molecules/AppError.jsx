import React from 'react';

const AppError = () => <p> There was an unexpected error </p>;

export default AppError;
