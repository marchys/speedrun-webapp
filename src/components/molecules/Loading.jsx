import React from 'react';
import CircularProgress from '@material-ui/core/CircularProgress';

const Loading = () => <CircularProgress size={50} />;

export default Loading;
