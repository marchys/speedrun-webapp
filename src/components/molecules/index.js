export { default as Loading } from './Loading';
export { default as AppError } from './AppError';
