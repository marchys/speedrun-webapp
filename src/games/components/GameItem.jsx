import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/fp/noop';

import ListItem from '@material-ui/core/ListItem';
import Avatar from '@material-ui/core/Avatar';
import ListItemText from '@material-ui/core/ListItemText';

const styles = {
  bigAvatar: {
    width: 80,
    height: 80,
  },
};

export default function GameItem({ name, image, onClick }) {
  return (
    <ListItem button onClick={onClick}>
      <Avatar src={image} alt={`${name} logo`} style={styles.bigAvatar} />
      <ListItemText primary={name} />
    </ListItem>
  );
}

GameItem.propTypes = {
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  onClick: PropTypes.func,
};

GameItem.defaultProps = {
  onClick: noop,
};
