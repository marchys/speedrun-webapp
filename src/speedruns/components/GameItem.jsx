import React from 'react';
import PropTypes from 'prop-types';

import ListItem from '@material-ui/core/ListItem';
import Avatar from '@material-ui/core/Avatar';
import ListItemText from '@material-ui/core/ListItemText';

const styles = {
  bigAvatar: {
    width: 80,
    height: 80,
  },
};

export default function GameItem({ name, image }) {
  return (
    <ListItem button>
      <Avatar src={image} alt={`${name} logo`} style={styles.bigAvatar} />
      <ListItemText primary={name} />
    </ListItem>
  );
}

GameItem.propTypes = {
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};
