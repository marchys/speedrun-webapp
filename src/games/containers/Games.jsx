import React from 'react';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import compose from 'lodash/fp/compose';

import List from '@material-ui/core/List';
import { Loading, AppError } from 'components/molecules';

import GameItem from 'games/components/GameItem';
import selectors from 'games/redux/selectors';
import actions from 'games/redux/actions';

const style = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
    alignItems: 'center',
  },
  list: {
    backgroundColor: '#68a88b',
    width: '100%',
  },
};

export class Games extends React.Component {
  componentDidMount() {
    const { getGames } = this.props;
    getGames();
  }

  renderGames() {
    const { games, loading, error } = this.props;

    if (loading) {
      return <Loading />;
    }

    if (error) {
      return <AppError />;
    }

    return (
      <List style={style.list}>
        {games.map(({ id, name, image }) => (
          <GameItem key={id} name={name} image={image} />
        ))}
      </List>
    );
  }

  render() {
    return (
      <div style={style.container}>
        <h1 style={style.title}> Games </h1>
        {this.renderGames()}
      </div>
    );
  }
}

Games.propTypes = {
  getGames: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  games: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
    }),
  ),
};

Games.defaultProps = {
  games: [],
  error: void 0,
};

const mapStateToProps = createStructuredSelector({
  loading: selectors.loading,
  games: selectors.games,
  error: selectors.error,
});

const connectToStore = connect(
  mapStateToProps,
  {
    getGames: actions.getGames,
  },
);

export default compose(connectToStore)(Games);
