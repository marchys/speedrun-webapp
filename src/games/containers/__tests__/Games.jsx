import React from 'react';
import { shallow } from 'enzyme';

import GameItem from 'games/components/GameItem';
import { Loading, AppError } from 'components/molecules';
import { Games } from '../Games';

const defaultGames = [
  { id: 'fe312', name: 'spiderman', image: 'https://www.google.com/' },
  { id: 'fasdtg', name: 'batman', image: 'https://github.com/' },
  { id: 'awetras', name: 'fallout', image: 'https://about.gitlab.com/' },
];

function setup({
  games = defaultGames,
  getGames = jest.fn(),
  loading = false,
  loaded = false,
  ...props
} = {}) {
  const wrapper = shallow(
    <Games games={games} getGames={getGames} loading={loading} loaded={loaded} {...props} />,
  );
  return {
    wrapper,
    get gameItems() {
      return wrapper.find(GameItem);
    },
    get title() {
      return wrapper.find('h1').text();
    },
    get loading() {
      return wrapper.find(Loading);
    },
    get error() {
      return wrapper.find(AppError);
    },
  };
}

test('should render', () => {
  const { wrapper } = setup();

  expect(wrapper.getElement()).not.toBe(null);
});

test('should show loading if loading prop', () => {
  const wrapper = setup({ loading: true });

  expect(wrapper.loading).toExist();
});

test('should show error if error prop', () => {
  const wrapper = setup({ error: { status: 500 } });

  expect(wrapper.error).toExist();
});

test('should render title', () => {
  const games = setup();

  expect(games.title).toMatchSnapshot();
});

test('should render a list of games passed', () => {
  const games = setup({ games: defaultGames });

  expect(games.gameItems).toHaveLength(3);
});
