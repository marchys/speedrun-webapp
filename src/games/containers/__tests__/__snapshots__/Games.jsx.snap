// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render title 1`] = `" Games "`;
